src/zenml/utils/source_utils.py

```diff
@@ -181,6 +190,94 @@ def get_module_source_from_class(
+    kernel environment, then this monkey-patches `inspect` module temporarily
+    with a workaround to get source from the cell.
+
+    Raises:
+        TypeError: If source not found.
+    """
+    # This is to be replaced with Environment().running_in_notebook.
+    def _running_in_notebook() -> bool:
+        """Detect whether we're running in a Jupyter notebook or not"""
+        try:
+            from IPython import get_ipython  # type: ignore
+
+            if get_ipython() is None:
+                # IPython is installed but not running from a notebook
+                return False
+            else:
+                return True
+        except ImportError:
+            # We do not even have IPython installed
+            return False
+
+    if _running_in_notebook():
+        # Monkey patch inspect.getfile temporarily to make getsource work.
+        # Source: https://stackoverflow.com/questions/51566497/
+        def _new_getfile(
+            object: Any,
+            _old_getfile: Callable[
+                [
+                    Union[
+                        ModuleType,
+                        Type[Any],
+                        MethodType,
+                        FunctionType,
+                        TracebackType,
+                        FrameType,
+                        CodeType,
+                        Callable[..., Any],
+                    ]
+                ],
+                str,
+            ] = inspect.getfile,
+        ) -> Any:
+            if not inspect.isclass(object):
+                return _old_getfile(object)
+
+            # Lookup by parent module (as in current inspect)
+            if hasattr(object, "__module__"):
+                object_ = sys.modules.get(object.__module__)
+                if hasattr(object_, "__file__"):
+                    return object_.__file__  # type: ignore[union-attr]
+
+            # If parent module is __main__, lookup by methods
+            for name, member in inspect.getmembers(object):
+                if (
+                    inspect.isfunction(member)
+                    and object.__qualname__ + "." + member.__name__
+                    == member.__qualname__
+                ):
+                    return inspect.getfile(member)
+            else:
+                raise TypeError("Source for {!r} not found".format(object))
+
+        # Monkey patch, compute source, then revert monkey patch.
+        _old_getfile = inspect.getfile
+        inspect.getfile = _new_getfile
+        src = inspect.getsource(value)
+        inspect.getfile = _old_getfile
+    else:
+        # Use standard inspect if running outside a notebook
+        src = inspect.getsource(value)
+    return src
+
+
+def get_hashed_source(value: Any) -> str:
+    """Returns a hash of the objects source code."""
+    try:
+        source_code = get_source(value)
+    except TypeError:
+        raise TypeError(
+            "Unable to compute the hash of the source code of the step. "
```

Comment:
I feel like this error message should be more generic now that the method is separated from any step/pipeline code and might be used in different scenarios